theatre-js/theatre
# Publish the development versions of the packages as tarballs

# Description

> :warning: We decided to test publishing the dev packages to npm instead of an object storage. This PR is likely to be closed without being merged!

This PR adds a script that generates tarballs from the `@theatre` packages and publishes them in the CI. [See this  (private) Notion doc](https://www.notion.so/theatrejs/POC-Publishing-development-packages-2e282feb3e85408aac9738eeb0a53e72) for the behind-the-scenes information.

# Versioning of the dev packages
A CI job builds the packages and assigns them a version number based on the version (without tags like `-dev.3`) specified in the root `package.json` (or `packages/r3f/package.json` in the case of `r3f` since that has an independent release schedule) and the abbreviated hash of the latest commit. Here's an example:
| prop | value |
| --- | --- |
| package name | `@theatre/core` |
| root version | `0.4.8-dev.3` |
| latest commit hash | `c8884093` |
| dev package version | `0.4.8-insiders.c8884093` |
| url (the domain will change to `packages.theatrejs.com`, or something similiar in the future) | `https://theatrejs-packages.nyc3.cdn.digitaloceanspaces.com/@theatre/core-c8884093.tgz` |
| install command | `yarn add https://theatrejs-packages.nyc3.cdn.digitaloceanspaces.com/@theatre/core-c8884093.tgz` 

# Testing this PR

1. Clone this repo: https://github.com/fulopkovacs/theatre-r3f-create-react-app
```
git clone git@github.com:fulopkovacs/theatre-r3f-create-react-app.git
```
2. Run the following commands in the repo (the second command takes quite a while):
```
yarn && yarn start
```
3. Visit `http://localhost:3000` in the browser
_Optional_:  You can verify that you're using the dev packages by checking the version of the `@theatre` packages with `yarn why <package-name>`, like:
```
$ yarn why @theatre/core
yarn why v1.22.10
[1/4] Why do we have the module "@theatre/core"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@theatre/core@0.4.8-insiders.c8884093"
info Has been hoisted to "@theatre/core"
info This module exists because it's specified in "dependencies".
info Disk size without dependencies: "656KB"
info Disk size with unique dependencies: "1.13MB"
info Disk size with transitive dependencies: "3.73MB"
info Number of shared dependencies: 2
Done in 0.42s.
```

This project wouldn't get built with the newest _published_ versions of the `@theatre` packages: you can try it in the same repo on the `broken` branch.

# :warning:  Limitations
CodeSandbox doesn't seem to support loading tarballs from urls:
- issue: https://github.com/codesandbox/codesandbox-client/issues/1570
- example for the error we get with CodeSandbox: [link](https://codesandbox.io/s/crazy-fire-6zro0w?file=/package.json:502-586)
 

# TODO
- [x] there's something weird around the commit hash in the CI (for some reason `git log -1` returns a different commit than the last one of this branch, so I'm guessing I have to use the env var provided by GH (`GITHUB_SHA`, [docs](https://docs.github.com/en/enterprise-server@3.4/actions/learn-github-actions/environment-variables#default-environment-variables)))
- [ ] (optional) check if publishing the dev packages can be extracted into a separate job (it would make our CI cleaner)
